larq_zoo/literature/real_to_bin_nets.py

```diff
@@ -0,0 +1,491 @@
+
+
+@factory
+class StrongBaselineNetBNN(StrongBaselineNetFactory):
+    model_name = Field("baseline_bnn")
+    input_quantizer = Field("ste_sign")
+    kernel_quantizer = Field("ste_sign")
+    kernel_constraint = Field("weight_clip")
+
+
+@factory
+class RealToBinNetFP(RealToBinNetFactory):
+    model_name = Field("r2b_fp")
+    input_quantizer = Field(lambda: tf.keras.layers.Activation("tanh"))
+    kernel_quantizer = Field(None)
+    kernel_constraint = Field(None)
+    kernel_regularizer = Field(lambda: tf.keras.regularizers.l2(1e-5))
+
+
+@factory
+class RealToBinNetBAN(RealToBinNetFactory):
+    model_name = Field("r2b_ban")
+    input_quantizer = Field("ste_sign")
+    kernel_quantizer = Field(None)
+    kernel_constraint = Field(None)
+    kernel_regularizer = Field(lambda: tf.keras.regularizers.l2(1e-5))
+
+
+@factory
+class RealToBinNetBNN(RealToBinNetFactory):
+    model_name = Field("r2b_bnn")
+    input_quantizer = Field("ste_sign")
+    kernel_quantizer = Field("ste_sign")
+    kernel_constraint = Field("weight_clip")
+
+
+@factory
+class ResNet18FP(ResNet18Factory):
+    model_name = Field("resnet_fp")
+    input_quantizer = Field(None)
+    kernel_quantizer = Field(None)
+    kernel_constraint = Field(None)
+
+
+def RealToBinaryNet(
+    *,  # Keyword arguments only
+    input_shape: Optional[Sequence[Optional[int]]] = None,
+    input_tensor: Optional[tf.Tensor] = None,
+    weights: Optional[str] = "imagenet",
+    include_top: bool = True,
+    num_classes: int = 1000,
+) -> tf.keras.models.Model:
+    """Instantiates the BNN version of the Real-to-Binary network from Martinez et. al.
+
+    Optionally loads weights pre-trained on ImageNet.
+
+    ```netron
+    real-to-bin-net-v0.1.0/r2b.json
+    ```
+    ```plot-altair
+    /plots/r2b_final_stage.vg.json
+    ```
+
+    # ImageNet Metrics
+
+    | Top-1 Accuracy | Top-5 Accuracy | Parameters | Memory  |
+    | -------------- | -------------- | ---------- | ------- |
+    | 65.01 %        | 85.72 %        | 11 995 624 | 5.13 MB |
+
+    # Arguments
+    input_shape: Optional shape tuple, to be specified if you would like to use a model
+        with an input image resolution that is not (224, 224, 3).
+        It should have exactly 3 inputs channels.
+    input_tensor: optional Keras tensor (i.e. output of `layers.Input()`) to use as
+        image input for the model.
+    weights: one of `None` (random initialization), "imagenet" (pre-training on
+        ImageNet), or the path to the weights file to be loaded.
+    include_top: whether to include the fully-connected layer at the top of the network.
+    num_classes: optional number of classes to classify images into, only to be
+        specified if `include_top` is True, and if no `weights` argument is specified.
+
+    # Returns
+    A Keras model instance.
+
+    # Raises
+    ValueError: in case of invalid argument for `weights`, or invalid input shape.
```

Comment:
Is it possible for a `NotImplementedError` to be raised?

Reply:
It was, fixed in d4b4c78